engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java

```diff
@@ -1,6 +1,27 @@
 package org.enso.interpreter.runtime.data;
 
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.EnsoLanguage;
 
 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+@ExportLibrary(InteropLibrary.class)
```

Comment:
> Let's see if Truffle API offers some better alternative to achieve such a default common implementation of some ExportLibrary messages?

I tried to leave `EnsoObject` as an interface, annotate it with `@ExportLibrary(InteropLibrary.class)`, and implement `hasLanguage` and `getLanguage` as default methods. But `@ExportMessage` on default method on interface does not work - the Truffle DSL processor fails with an error. Any ideas are welcome.

As for `toDisplayString` implementation - that is required by Truffle DSL, because if you export one of `getLanguage` or `hasLanguage` messages, then, `toDisplayString` is abstract and has to be implemented. However, I am still not sure what is the proper way to define it here, in `EnsoObject`.

Reply:
I tried to use `@ExportLibrary(receiverType=EnsoObject.class)` but no luck. The best I could get with having an interface and also `Base` class is:
```diff
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java
new file mode 100644
index 0000000000..2545ad6dbe
--- /dev/null
+++ engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java
@@ -0,0 +1,15 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.runtime.data.EnsoObject;
+
+@ExportLibrary(InteropLibrary.class)
+class TFO extends EnsoObject.Base implements EnsoObject {
+    @ExportMessage
+    @Override
+    public final String toDisplayString(boolean internal) {
+        return "TFO";
+    }
+}
diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
index ce60158116..f569c98256 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
@@ -1,6 +1,24 @@
 package org.enso.interpreter.runtime.data;
 
+import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.EnsoLanguage;
 
 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+public interface EnsoObject extends TruffleObject {
+    @ExportLibrary(InteropLibrary.class)
+    public static abstract class Base implements EnsoObject {
+        @ExportMessage
+        public final boolean hasLanguage() {
+            return true;
+        }
+        @ExportMessage
+        public Class<? extends EnsoLanguage> getLanguage() {
+            return EnsoLanguage.class;
+        }
+        @ExportMessage
+        public abstract String toDisplayString(boolean internal);
+    }
+}
```

Reply:
According to @chumer, we should go with the `abstract class` - the `instanceof` checks will be even faster than with interfaces.